File: tests/run-pass/rosetta/001_100_doors/lua/100_doors_profile.lua
-- 100 Doors -- characterization mirror of ./100_doors.lua.
--
-- Runs the SAME algorithm as simulate() over a reduced iteration count with a
-- counter incremented at every characterized operation site, and emits one
-- `work <category> <count>` line per counter. The harness runs this program in
-- its own untimed invocation, so the counters never enter a timed measurement.
--
-- The categories are the harness's FIXED language-independent vocabulary. Every
-- counter must equal its Ori and Python sibling exactly; the site list is
-- documented once in ../ori/100_doors_profile.ori.
--
-- Lua tables are 1-based, so Python's 0-based `range(p - 1, 100, p)` is written
-- here as `for idx = p, 100, p` and `doors[s]` as `doors[s + 1]`.

local ITERATIONS = 500

local function simulate_profiled(salt)
  local allocs = 0
  local ariths = 0
  local branches = 0
  local indexes = 0
  local iters = 0

  allocs = allocs + 1
  local doors = {}
  for i = 1, 100 do
    doors[i] = false
  end

  for p = 1, 100 do
    iters = iters + 1
    ariths = ariths + 1
    for idx = p, 100, p do
      iters = iters + 1
      indexes = indexes + 2
      doors[idx] = not doors[idx]
    end
  end

  ariths = ariths + 1
  local s = salt % 100
  indexes = indexes + 2
  doors[s + 1] = not doors[s + 1]

  local open = 0
  for i = 1, 100 do
    iters = iters + 1
    indexes = indexes + 1
    branches = branches + 1
    if doors[i] then
      ariths = ariths + 1
      open = open + 1
    end
  end

  return allocs, ariths, branches, indexes, iters, open
end

local function main()
  local calls = 0
  local allocs = 0
  local ariths = 0
  local branches = 0
  local indexes = 0
  local iters = 0
  local checksum = 0
  for n = 0, ITERATIONS - 1 do
    local a, ar, b, ix, it, c = simulate_profiled(n)
    calls = calls + 1
    allocs = allocs + a
    ariths = ariths + ar
    branches = branches + b
    indexes = indexes + ix
    iters = iters + it
    checksum = checksum + c
  end

  print(string.format("work alloc %d", allocs))
  print(string.format("work arith %d", ariths))
  print(string.format("work branch %d", branches))
  print(string.format("work call %d", calls))
  print("work field 0")
  print(string.format("work index %d", indexes))
  print(string.format("work loop_iter %d", iters))
  print("work string_op 0")
  print("work call_sites 1")
  print("work call_targets 1")
end

main()

File: tests/run-pass/rosetta/001_100_doors/python/100_doors_profile.py
# 100 Doors -- characterization mirror of ./100_doors.py.
#
# Runs the SAME algorithm as simulate() over a reduced iteration count with a
# counter incremented at every characterized operation site, and emits one
# `work <category> <count>` line per counter. The harness runs this program in
# its own untimed invocation, so the counters never enter a timed measurement.
#
# The categories are the harness's FIXED language-independent vocabulary. Every
# counter must equal its Ori and LuaJIT sibling exactly; the site list is
# documented once in ../ori/100_doors_profile.ori.

ITERATIONS = 500


def simulate_profiled(salt: int) -> tuple[int, int, int, int, int, int]:
    allocs = 0
    ariths = 0
    branches = 0
    indexes = 0
    iters = 0

    allocs += 1
    doors = [False] * 100

    for p in range(1, 101):
        iters += 1
        ariths += 1
        for idx in range(p - 1, 100, p):
            iters += 1
            indexes += 2
            doors[idx] = not doors[idx]

    ariths += 1
    s = salt % 100
    indexes += 2
    doors[s] = not doors[s]

    open_count = 0
    for i in range(100):
        iters += 1
        indexes += 1
        branches += 1
        if doors[i]:
            ariths += 1
            open_count += 1

    return allocs, ariths, branches, indexes, iters, open_count


def main() -> None:
    calls = 0
    allocs = 0
    ariths = 0
    branches = 0
    indexes = 0
    iters = 0
    checksum = 0
    for n in range(ITERATIONS):
        a, ar, b, ix, it, c = simulate_profiled(n)
        calls += 1
        allocs += a
        ariths += ar
        branches += b
        indexes += ix
        iters += it
        checksum += c

    print(f"work alloc {allocs}")
    print(f"work arith {ariths}")
    print(f"work branch {branches}")
    print(f"work call {calls}")
    print("work field 0")
    print(f"work index {indexes}")
    print(f"work loop_iter {iters}")
    print("work string_op 0")
    print("work call_sites 1")
    print("work call_targets 1")


main()
